sequence_task_3/student/include/impl.h

```diff
@@ -7,46 +7,102 @@ template<class Container, class Song_t>
 class DynamicPlaylist
 {
 public:
-    /** @todo Member traits */
+    /** Member traits */
+    using value_type = typename Container::value_type;
+    using reference = typename Container::reference;
+    using const_reference = typename Container::const_reference;
+    using iterator = typename Container::iterator;
+    using const_iterator = typename Container::const_iterator;
+    using difference_type = typename Container::difference_type;
+    using size_type = typename Container::size_type;
+    using self_type = DynamicPlaylist<Container, Song_t>;
 
-    /** @todo Iterators */
+    /** Iterators */
+    const_iterator begin() {
+        return m_tracklist.cbegin();
+    }
+
+    const_iterator end() {
+        return m_tracklist.cend();
+    }
 
     DynamicPlaylist() = default;
+    DynamicPlaylist(const self_type& other) = default;
+    DynamicPlaylist(self_type&& other) = default;
+    self_type& operator=(const self_type& rhs) = default;
+    self_type& operator=(self_type&& rhs) = default;
 
-    /** @todo Constructor from two iterators*/
+    /** Constructor from two iterators*/
+    template <class Iter>
+    DynamicPlaylist(Iter iterBegin, Iter iterEnd)
+         : m_tracklist(iterBegin, iterEnd)
+     {}
 
-    /** @todo Set track as currently played from initializer */
+    /** Set track as currently played from initializer */
     template<class... Args>
-    const Song_t& play(Args&&... songData);
+    const Song_t& play(Args&&... songData) {
+        m_tracklist.emplace_front(std::forward<Args>(songData)...);
+        return current();
+    }
 
-    /** @todo Set track as currently played from lvalue */
-    const Song_t& play(const Song_t& song);
+    /** Set track as currently played from lvalue */
+    const Song_t& play(const Song_t& song) {
+        m_tracklist.push_front(song);
+        return current();
+    }
 
-    /** @todo Set track to be played next from initializer */
+    /** Set track to be played next from initializer */
     template<class... Args>
-    void playNext(Args&&... songData);
+    void playNext(Args&&... songData) {
+        if (!hasTracks()) {
+            play(std::forward<Args>(songData)...);
+            return;
+        }
+        m_tracklist.emplace(std::next(m_tracklist.begin()), std::forward<Args>(songData)...);
+    }
 
-    /** @todo Set track to be played next from lvalue */
-    void playNext(const Song_t& song);
+    /** Set track to be played next from lvalue */
+    void playNext(const Song_t& song) {
+        if (!hasTracks()) {
+            play(song);
+            return;
+        }
+        m_tracklist.push_front(song);
```

Comment:
please check the difference with line 61 (insert position)

Reply:
Done!